### ISSUE TITLE
SEC-2489: Document equals and hashcode should be overridden on UserDetails when using concurrent session authentication control

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-security#2709]
[Quinten Krijger](https://jira.spring.io/secure/ViewProfile.jspa?name=qkrijger) (Migrated from [SEC-2489](https://jira.spring.io/browse/SEC-2489?redirect=false)) said:

The implementation of the ConcurrentSessionControlAuthenticationStrategy calls SessionRegistryImpl.getAllSessions, which uses a map from principal to sessions. Therefore, if one implements UserDetails the equals() and hashcode() should be overridden. Otherwise, the strategy will not work.

My proposal here is to document a warning at http://docs.spring.io/spring-security/site/docs/3.2.x/reference/htmlsingle/#ns-concurrent-sessions and http://docs.spring.io/spring-security/site/docs/3.2.x/reference/htmlsingle/#concurrent-sessions that the UserDetails implementation should override equals and hashcode in order for the stategy to function.

**Comment by ledsoft** [comment]
:+1: 
Agreed. Just spent several hours trying to figure out why I am able to login twice into my application when maxSessions is set to 1.

**Comment by Longwater1234** [comment]
Yes. Please document this in your official Spring reference website. I suggest adding it in highlighted **NOTE** text box.
